### PR TITLE
MAINT: optimize: refactor common linprog arguments into namedtuple

### PR DESCRIPTION
**Pull request by zchelseal**
#### Reference issue
Partially addresses #9671

#### What does this implement/fix?
It refactors `c, A_ub, b_ub, A_eq, b_eq, bounds, x0` into a `namedtuple`.

#### Summary
- Added the proposed `namedtulple` into `_linprog_util.py` 
- Refactored related functions in `_linprog_util.py`, `_linprog.py`, and tests for `_clean_inputs`
- Modified assertion on bounds in `test_aliasing` and `test_aliasing2` (`np.inf` -> `None` and `-np.inf` -> `None`)

### CODE REVIEW

**Comment by mdhaber** [comment]
@Kai-Striega does this look like what you had in mind in #9671?

File: scipy/optimize/_linprog_util.py
Comment (Kai-Striega) [inline scipy/optimize/_linprog_util.py]: `c`?

File: scipy/optimize/_linprog_util.py
Comment (Kai-Striega) [inline scipy/optimize/_linprog_util.py]: Good use of `.rjust` :+1: Maybe keep ``self.__class__.__name__`` before both outputs?

**Review by Kai-Striega — COMMENTED** [review]
Looks really good to me. Correct Documentation and good use of dunder methods throughout.  Well done @zchelseal

File: scipy/optimize/_linprog_util.py
Comment (zchelseal) [reply scipy/optimize/_linprog_util.py]: `c` was left out intentionally as it's the only mandatory argument for an LP - this way if `c` is called without being initialized with a value first, it would throw an `AttributeError` :)
Let me know if this is not the expected behaviour xD

File: scipy/optimize/_linprog_util.py
Comment (zchelseal) [reply scipy/optimize/_linprog_util.py]: Will do! Thanks for the suggestion 👍

File: scipy/optimize/_linprog_util.py
Comment (mdhaber) [reply scipy/optimize/_linprog_util.py]: Maybe accept `c` as a parameter (independent of `args`/`kwargs`) with no default value?

Comment (zchelseal) [reply scipy/optimize/_linprog_util.py]: Would this suffice?
```Python
def __init__(self, c, *args, **kwargs):
    for key in ['A_ub', 'b_ub', 'A_eq', 'b_eq', 'bounds', 'x0']:
        self[key] = None
    dict.__init__(self, *args, **dict(kwargs, c=c))
```
Nitpick: the difference in initialization (compared to a regular dict) might be worth documenting:
```Python
lp1 = LPProblem(c=[-1, 4], A_ub=[[-3, 1], [1, 2]], b_ub=[6, 4])  # probably the most used and readable approach
lp2 = LPProblem({'c': [-1, 4], 'A_ub': [[-3, 1], [1, 2]], 'b_ub': [6, 4]})  # unusable; lp2.c becomes the entire dictionary
lp3 = LPProblem([-1, 4], {'A_ub': [[-3, 1], [1, 2]], 'b_ub': [6, 4]})  # fine
lp4 = LPProblem([-1, 4], {'c': [2, 3], 'A_ub': [[-3, 1], [1, 2]], 'b_ub': [6, 4]})  # lp4.c = [-1, 4] not [2, 3]
```
I can add example usage in the comment string if we go with this approach :)

Comment (mdhaber) [reply scipy/optimize/_linprog_util.py]: `lp2`'s definition can work if you check whether the first parameter is a dictionary or not. If it is, you can assume it contains `c`, etc...
I don't think we should encourage definitions like `lp3` and `lp4`. 

Also, we didn't really discuss whether this class would be public or not, but it sounds like we're assuming it will be. I think that's fine, so users would then be able to define an `LPProblem` and use `linprog` to solve it:
"""
lp = LPProblem(...)
linprog(lp)
"""
but we will still need to allow the original syntax for calling `linprog`.

But I think @Kai-Striega's original purpose was to simplify passing all the information around _internally_.

Comment (zchelseal) [reply scipy/optimize/_linprog_util.py]: > The `lp2` initialization can work if you check whether the first parameter of `__init__` is a dictionary or not. If not, then the first argument is `c`. If it is, assume it contains the entire problem statement (including constraints, bounds, and guess).
I would advise against performing this check to avoid ambiguity in the variable `c` and type-checking in general. Instead I can add a note/ example usage in the docstring saying not to pass in a whole dictionary in a `lp2`-manner.

> I don't think we should encourage definitions like `lp3` and `lp4`.
Agreed! Was just trying to list out the corner cases :)

I think we should keep this class internal-use-only (and keep the parameter-passing for public methods as-is) to stick with the original motivation xD

Comment (Kai-Striega) [reply scipy/optimize/_linprog_util.py]: Another case, that currently works, is to specify values as arguments, e.g.:

```python
lp5 = LPProblem([-1, 4], [[-3, 1], [1, 2]], [6, 4])
```

If we do change publically this way would need to be doable.

**Comment by Kai-Striega** [comment]
@zchelseal @mdhaber My thoughts on being public/private

> But I think @Kai-Striega's original purpose was to simplify passing all the information around internally. 

Yes, that was the idea.

> Also, we didn't really discuss whether this class would be public or not, but it sounds like we're assuming it will be.

I'm not certain that it should be public (yet). We're already discussed other API changes e.g.  #10136 (Yes, I will get back to it.) or having array_like types for bounds. Maybe we could use it internally first, see how it fits, then make the major API changes within a single release?

**Comment by mdhaber** [comment]
That sounds good to me. 
Internally, I'd prefer to initialize the objects with separate arguments, no dictionary.

File: scipy/optimize/_linprog.py
Comment (mdhaber) [inline scipy/optimize/_linprog.py]: Can you explain how you got around making copies here? Were they not necessary?

File: scipy/optimize/_linprog.py
Comment (mdhaber) [inline scipy/optimize/_linprog.py]: Maybe put the whole `lp_o` in and unpack it later?

File: scipy/optimize/_linprog_util.py
Comment (mdhaber) [inline scipy/optimize/_linprog_util.py]: Cool I guess this is what @Kai-Striega suggested. I've never used a `nametuple` before.

**Review by mdhaber — COMMENTED** [review]
Didn't make it all the way through. Some thoughts to start.

**Comment by zchelseall** [comment]
> Didn't make it all the way through. Some thoughts to start.

Thank you so much @mdhaber for your speedy response! I was able to fix most of the affected test cases offline except 6, where I was a little lost and hoped to get your insights on :)

**Comment by zchelseal** [comment]
> Didn't make it all the way through. Some thoughts to start.

Thank you so much @mdhaber for your speedy response!
I was able to fix most of the affected test cases offline except 6, where I was a little lost and hoped to get your insights on :)

File: scipy/optimize/_linprog_util.py
Comment (zchelseal) [reply scipy/optimize/_linprog_util.py]: Me neither - I began realizing the 2 benefits of using a `namedtuple` as I started refactoring the the linprog modules using `_LPProblem` :)

1. Because the elements of a `namedtuple` are ordered, it can be easily initialized with `lp5 = LPProblem([-1, 4], [[-3, 1], [1, 2]], [6, 4])`, and unpacked with `c, A_ub, b_ub, A_eq, b_eq, bounds, x0 = lp` (whereas neither are possible with a dictionary).
2. `namedtuple`s are immutable, so instead of being able to modify element values in place, we always pass back a new `_LPProblem` so it's more explicit and readable for debugging.

File: scipy/optimize/_linprog.py
Comment (zchelseal) [reply scipy/optimize/_linprog.py]: Thanks for the reminder! It took me a while to figure out the best way to do this, but as it turns out `lp_o` is a copy of `lp` because of the immutability property of `namedtuple`.
Instead of having functions modify the value of `c`, `A_ub`, ... etc. in place, I do a `return _LPProblem(c, A_ub, b_ub, A_eq, b_eq, bounds, x0)` so that if debugging is needed in the future, it's easy to compare the before-and-after of elements to spot changes.

I'm suspecting though that I may have missed some of the implicit modifications that could have caused the 6 remaining failed tests.

Comment (mdhaber) [reply scipy/optimize/_linprog.py]: Not exactly. Even if you were to copy the tuple, it doesn't copy the objects the tuple refers to. The tuple only stores references to the objects, but both copies of the tuple would refer to the same objects because you've only copied the references.

Comment (zchelseal) [reply scipy/optimize/_linprog.py]: My apologies for misusing the word "copy"!
What I meant was that `lp_o` is intended to be a preserved reference of `lp` before `lp` gets re-assigned to a new `namedtuple` after `_presolve` (line 515).

```Python
lp_o = lp

# Solve trivial problem, eliminate variables, tighten bounds, etc...
c0 = 0  # we might get a constant term in the objective
if solver_options.pop('presolve', True):
    rr = solver_options.pop('rr', True)
    (lp, c0, x, undo, complete, status, message) = _presolve(lp, rr, tol) # here lp points to a new namedtuple
```

File: scipy/optimize/_linprog.py
Comment (zchelseal) [reply scipy/optimize/_linprog.py]: I was thinking about this as well for consistency, but I wasn't sure if it's a good idea to make `postsolve_args` nested, ie. 
```Python
# definition:
postsolve_args = (lp_o, undo)

# unpacking:
lp, undo = postsolve_args
c, A_ub, b_ub, A_eq, b_eq, bounds, x0 = lp
```
Or to pack `undo` into lp_o, ie.
```Python
# definition:
postsolve_args = lp_o._replace(undo=undo)

# unpacking:
c, A_ub, b_ub, A_eq, b_eq, bounds, x0, undo = postsolve_args
```
Would you prefer one way over another?

Comment (mdhaber) [reply scipy/optimize/_linprog.py]: I prefer the former. I don't have a problem with the double nesting.

I don't think `undo` belongs in `lp_o`. (To be honest, I'm not really sure that `x0` belongs inside it, but I wouldn't change it right now.)

**Comment by mdhaber** [comment]
@zchelseal Thanks for your patience! 
This got pretty big, so it will really take some time to sit down and review as it is. I'm thinking about whether there is a small adjustment we can make that will reduce the number of line changes. For instance, if `_clean_inputs` were to accept not only the new `lp` structure but also the old inputs, then we could revert many of the changes to the tests. That would make me a bit more comfortable about this, even though many of the changes are minor. 
What do you think, @Kai-Striega ? Is this what you envisioned?

File: scipy/optimize/_linprog_util.py
Comment (Kai-Striega) [inline scipy/optimize/_linprog_util.py]: :+1: The docstring looks really good. Maybe also show an example setting `A_eq`, `b_eq` without `A_ub`, `b_ub`? A namedtuple is still a tuple, so it may confuse some users who have not used them before that it's possible to initialise a tuple with keywords.

File: scipy/optimize/_linprog_util.py
Comment (Kai-Striega) [inline scipy/optimize/_linprog_util.py]: Perhaps it would be better to pass ``lp`` to ``_check_sparse_inputs`` directly? That way we won't have to think about explicitly replacing `A_ub`/`A_eq`. So  it would become:
```python
lp, solver_options = _check_sparse_inputs(solver_options, lp)
lp = _clean_inputs(lp)
return lp, solver_options
```
Looking at it now, if we change it, `lp` would need to be more descriptive

**Review by Kai-Striega — COMMENTED** [review]
> This got pretty big, so it will really take some time to sit down and review as it is.

Agreed. I've had a good look through it and it looks good from my perspective.

>  I'm thinking about whether there is a small adjustment we can make that will reduce the number of line changes. For instance, if _clean_inputs were to accept not only the new lp structure but also the old inputs, then we could revert many of the changes to the tests. That would make me a bit more comfortable about this, even though many of the changes are minor.

I second your sentiment that we need to be careful with such large changes.

If we are going to use_LPProblem, or similar, I would prefer to use it consistently, even for tests.  Personally, ``assert_allclose(lp_cleaned.c, np.array([1, 2]))`` seems *safer* than ``assert_allclose(outputs[0], np.array([1, 2]))`` as ``c`` is clearly specified.

> What do you think, @Kai-Striega?

I think it's good. In addition to the above comments, I've left some comments about specific things as review comments.

> Is this what you envisioned?

I think it's more comprehensive than what I had in mind, but I think that's a good thing. Overall good job @zchelseal.

**Comment by Kai-Striega** [comment]
@mdhaber would you have the time to give this a thorough read through? I don't think there is anything to stop us merging the PR as @zchelseal work looks good.

I can fix the merge conflicts if you think it's OK.

**Comment by mdhaber** [comment]
Thanks for your patience. I'm pretty backlogged. You had some comments - did you think they needed to be fixed before merge or were you happy with it anyway? 
If you think it's ready, other than merge conflicts, I'll try to take a look this weekend.

**Review by zchelseal — COMMENTED** [review]
> I think it's good. In addition to the above comments, I've left some comments about specific things as review comments.

Thank you both for the comments and my apologies for the late response!
I've been AFK for a couple of weeks since school started and trying to figure out a way to set up SciPy on my personal PC or a school Mac - I'll try to address these comments by the end of this week :)

**Comment by Kai-Striega** [comment]
@mdhaber I had some comments but they shouldn't stop us merging. At this point, the major blocker was discussing the test changes. 

> trying to figure out a way to set up SciPy on my personal PC or a school Mac -- @zchelseal 

What kind of computer are you using? I've been 50:50 about creating some tutorials on how I build it on Windows and Linux.

**Comment by mdhaber** [comment]
Is there a problem with the [Development Environment Quickstart Guides](http://scipy.github.io/devdocs/dev/contributor/contributor_toc.html#development-environment)?

@zchelseal I suggest Mac, but if you only have Windows, you can use the Docker instructions.

**Comment by Kai-Striega** [comment]
@mdhaber they are very good 👍. What I've wanted to add is a guide without using Anaconda and avoiding homebrew in favour of `apt` (that's already used to install git) and possibly an instruction video to compilment your Mac videos.

I've also been trying to build SciPy with Intel + MKL on Windows, but thats proving to be a challenge.

File: scipy/optimize/_linprog_util.py
Comment (person142) [reply scipy/optimize/_linprog_util.py]: Apologies for the drive-by review, but now that we require Python >= 3.5 I believe we can use the newer syntax for named tuples:

```
import typing as t

class _LPProblem(t.NamedTuple):
    “””docstring looks nicer now”””

    c: <type-of-c>
    A_ub: ...
```

**Comment by mdhaber** [comment]
@kai-striega I agree. That cuts down on the number of instructions for sure. I've wanted to add that too, but first I wanted to fine out how these should all be organized. There are so many combinations of options for doing the same things. Can we set these up so they are easier to maintain than a separate file for each combination? 

If you can help me merge #10461 I can help you with non-Anaconda guides.

**Comment by Kai-Striega** [comment]
Sure thing. It's nearly midnight here, but, I'll take a look tomorrow.

**Review by mdhaber — CHANGES_REQUESTED** [review]
This does clean up the appearance of some function signatures.
I have not verified that every letter of every copy of the docstring is correct, nor that all aspects of every test were modified in a way that is faithful to the original; I have to trust that the refactoring was performed carefully.
I did notice that `test_aliasing` and `test_aliasing2` were modified in a way that does _not_ preserve the original intent. Please revert and re-refactor to preserve the original intent.
Also, please see whether the regular tuple in line 1267 should be a `_LPProblem`.

File: scipy/optimize/_linprog_util.py
Comment (Kai-Striega) [reply scipy/optimize/_linprog_util.py]: @person142 I think having the typing would be nice. Now that I think about it, should this be consistent throughout the remaining linprog code (but as a new PR). @mdhaber thoughts?

Comment (zchelseal) [reply scipy/optimize/_linprog_util.py]: `_LPProblem` would definitely look nicer using `typing`, but I wonder if we should switch to this syntax after we require Python >= 3.6, because [variable annotation syntax](https://www.python.org/dev/peps/pep-0526/#global-and-local-variable-annotations) was added in version 3.6 (see [typing.NamedTuple](https://docs.python.org/3.6/library/typing.html#typing.NamedTuple)).

This would affect how we make `c` the only required arg and set default values to `None` for `A_ub`, `b_ub`, etc.
Currently done this way:
```Python
_LPProblem.__new__.__defaults__ = (None,) * 6  
```
After version 3.6:
```
class _LPProblem(t.NamedTuple):
    “””docstring”””

    c: <type-of-c>  # only required arg
    A_ub: <type-of-A_ub> = None  # default to None
    b_ub: <type-of-b_ub> = None  # default to None
    ...
```

There may be a way to do this within class definition before version 3.6 too - if this is the case please let me know and I can add it :)

**Comment by mdhaber** [comment]
Oops. This sort of thing has happened to me before, but I'm not an expert at diagnosing and fixing it yet. @zchelseal do you know what caused all these other things to appear? @Kai-Striega suggestions?

**Comment by zchelseal** [comment]
@mdhaber This seems to be unintended behaviour caused by GitHub not tracking new commits to the target branch since PR opened - I found a workaround [here](https://stackoverflow.com/questions/16306012/github-pull-request-showing-commits-that-are-already-in-target-branch) and it should be fixed now!

**Comment by rgommers** [comment]
Hmm weird failure for the `pyflakes` check. I re-ran the CI job, it's real:
```
scipy/optimize/tests/test__linprog_clean_inputs.py

The command "grep -rlIP '[^\x00-\x7F]' scipy | grep '\.pyx\?' | sort > unicode.out; grep -rlI '# -\*- coding: \(utf-8\|latin-1\) -\*-' scipy | grep '\.pyx\?' | sort > coding.out; comm -23 unicode.out coding.out > test_code.out; cat test_code.out;  test \! -s test_code.out

" exited with 1.
```

Unlike normal style checker failures, it doesn't tell you what's wrong. Perhaps you can reproduce locally?

File: scipy/optimize/tests/test__linprog_clean_inputs.py
Comment (WarrenWeckesser) [inline scipy/optimize/tests/test__linprog_clean_inputs.py]: The blank character that separates `'a'` and `'side'` is a "no-break space".  That's Unicode code point U+00A0, and encoded in UTF-8, it is the two bytes `b'\xc2\xa0'`.  This should be changed to an ASCII space.

File: scipy/optimize/_linprog_util.py
Comment (person142) [reply scipy/optimize/_linprog_util.py]: @zchelseal ach you are right-I forgot that the variable annotations got added later!

File: scipy/optimize/_linprog_util.py
Comment (WarrenWeckesser) [inline scipy/optimize/_linprog_util.py]: Make the number of dashes match the length of the title above them.

**Comment by WarrenWeckesser** [comment]
I added a couple drive-by comments (not a full review).

**Comment by zchelseal** [comment]
@mdhaber Would you mind taking another look for any other changes I should make before merging? 
I've fixed the `test_aliasing`s to check for in-place modifications along with the other suggested changes (thanks to everyone who reviewed!) :)

**Comment by mdhaber** [comment]
I will try, but I'm traveling internationally this week so I'm not sure when I can. Thank you for working on this!

File: scipy/optimize/tests/test__linprog_clean_inputs.py
Comment (mdhaber) [reply scipy/optimize/tests/test__linprog_clean_inputs.py]: In the future, please leave comments like this unresolved. The reviewer should determine whether they are resolved or not. Thanks.

File: scipy/optimize/_linprog_util.py
Comment (mdhaber) [reply scipy/optimize/_linprog_util.py]: @Kai-Striega Is this resolved?

**Review by mdhaber — COMMENTED** [review]
Please try to remove the changes to scipy-sphinx-theme, @zchelseal, then I think that's it from me. I have read through this a few times and have commented on what I have seen, but I think this needs another review from @Kai-Striega, then I will let him decide on it. 

We should try to merge this soon, though, so that we have a lot of time before the next release to work with the refactored code to find any subtle issues not caught by the review. 

In the meantime, @zchelseal, have you thought about what you'd like to tackle next? We'd love it if you could continue to help with `linprog`.

**Comment by tylerjereddy** [comment]
Looks like I will have to bump this to 1.5.0, but it certainly looks like some high quality work is happening here.

I bet those sphinx-theme changes just snuck in with something like `git add -u` after switching branches. I catch myself doing that sometimes..

**Review by Kai-Striega — COMMENTED** [review]
This PR has expanded significantly (for the better) that what I had had in mind when making the original request. @zchelseal's work looks to be of a really high calibre and think the contribution will be a net positive. Both Matt and I seem struggle with the scope of the changes, making a proper review difficult in addition to having had several months since last reading through the changes.

Saying that, I approved my last review (in August) and Matt's review was positive modulo some changes which have now been fixed. Further the test suite passes and I have experimented with the changes local,  all of which return the expected result.

I agree with @mdhaber's view that if we are to merge the best time would be now, giving us as much time as possible to identify and fix changes before the 1.6 release.

**Comment by Kai-Striega** [comment]
@mdhaber you still have some "changes requested" are you satisfied that those changes in particular have been fixed?

@zchelseal There are some sphinx-related changes which I would like to avoid merging, do you have any ideas for how to remove them?

**Review by mdhaber — APPROVED** [review]

**Comment by mdhaber** [comment]
Well, I just approved the changes to not hold up your merge, but I think my "changes requested" was for removing the sphinx changes.

**Comment by Kai-Striega** [comment]
@mdhaber or @tylerjereddy there have recently been some new issues discussing possible enhancements to linprog. I would like to try to get this merged ASAP otherwise I think it will lead to major merge conflicts later. 

Is there a way for me to "nicely" remove the sphinx changes myself while still giving the commit credit to @zchelseal?

**Comment by rgommers** [comment]
> Is there a way for me to "nicely" remove the sphinx changes myself while still giving the commit credit to @zchelseal?

@Kai-Striega authorship is per commit; you can just push a new commit to `zchelseal:refactor_linprog_arguments` that undoes the unwanted changes, and everything will be as it should be.

**Comment by Kai-Striega** [comment]
@rgommers thanks! I've fixed the merge conflicts - I'll let the CI run just to make sure it is all done properly then remove the sphinx changes then (hopefully) merge.

**Comment by mdhaber** [comment]
Thanks @zchelseal! If you'd like your name added to [THANKS.txt](https://github.com/scipy/scipy/blob/master/THANKS.txt), please make an entry [here](https://github.com/scipy/scipy/wiki/*THANKS.txt-additions-modifications-for-1.5.0*).

**Comment by mdhaber** [comment]
Hi @zchelseal, just thought I'd thank you again and invite you back! Let me know if you're looking for something interesting to tinker with; there are always things to do!

**Comment by zchelseal** [comment]
@mdhaber Thank you so much for the message and my apologies for the late response! Things have been a little overwhelming lately :'( 
I'd love to work on more issues in `scipy.optimize` after my current internship - is there something I can help on starting end of May?

**Comment by Kai-Striega** [comment]
@zchelseal great to hear that you are still interested.  Matt has kept track of additional cleanups (#9671) and a wishlist (#9269) for ``linprog`` - if you're looking for more issues that's the best place to start.

Also, congratulations on getting an internship!

**Comment by mdhaber** [comment]
Yeah, anything on those lists. I think these two would be pretty easy:
* allow bounds to be specified as `Bounds` objects (#10288)
* automatically remove constraints with infinite upper bound (#10288)

and anything to do with sensitivity analysis. We're in the process of adding a new solver and I think those things would still be useful afterwards.

**Pull request merged**
